doc/man3/OSSL_HPKE_CTX_new.pod

```diff
@@ -0,0 +1,328 @@
+
+OSSL_HPKE_recommend_ikmelen() allows the caller to get a recommendation
+for how long (in octets) an IKM to use with a given suite. This is needed
+in cases where the same public value needs to be (re-)generated by a 
+sender before calling OSSL_HPKE_seal().
+
+OSSL_HPKE_get_grease_value() produces values of the appropriate length (for the given
+suite) so that a protocol using HPKE can send so-called GREASE (see RFC8701) values
+that are harder to distinguish from a real use of HPKE.
+
+OSSL_HPKE_str2suite() maps known strings HPKE cipher suites. String values are
+defined in hpke.h. The suite strings recognised are comma-separated KEM, KDF
+and AEAD names in that order, e.g. "x25519,hkdf-sha256,aes128gcm". This can
+be used by e.g., command line tools that accept string form names for 
+HPKE codepoints.
+
+=head2 Key Generation
+
+Any OpenSSL-supported method producing an EVP_PKEY for generating keys that
+match the IANA-registered codepoints for HPKE may be used. We also provide
+a specific API that is aware of how those codepoints map to OpenSSL algorithm
+implementations.
+
+OSSL_HPKE_keygen() generates an ECDH private value for use with the
+given HPKE suite. This returns the private key as an
+EVP_PKEY*. Only the kem_id field of the suite input is currently used. If the
+initial key material (IKM) inputs (ikmlen/ikm) are (0/NULL) then a key pair
+will be randomly generated. If those input are nonzero/non-NULL then the 
+deterministic key generation scheme from RFC9180 will be used.
+
+=head2 OSSL_HPKE_CTX manipulating APIs
+
+Callers can construct an OSSL_HPKE_CTX object using OSSL_HPKE_CTX_new() 
+that removes the need to provide repeated input parameters such as the
+libctx, propq, HPKE mode and suite. The OSSL_HPKE_CTX_free()
+function releases that memory.
+
+Three additional attributes may be set for such context objects: a 
+pre-shared key (PSK), a sender authentication key and to control (via set/get)
+the sequence number for the next seal/open operation.
+
+Sender key generation can also be based on the deterministic key generation
+scheme specified in RFC 9180 via the OSSL_HPKE_CTX_set1_ikme() function.
+
+If the "EXPORT-ONLY" mode from RFC 9180 is desired, then OSSL_HPKE_export()
+can be used after a call to OSSL_HPKE_sender_export_encap() or
+OSSL_HPKE_recipient_export_decap() if the relevant export-only AEAD has been
+chosen (OSSL_HPKE_AEAD_ID_EXPORTONLY).
+
+=head1 RETURN VALUES
+
+All functions return 1 for success. Functions may return a negative 
+number, zero or a positive number in the event of an error.
+
+=head1 EXAMPLES
+
+This example demonstrates a minimal round-trip using HPKE.
+
+    #include <openssl/hpke.h>
+    /* we'll do a round-trip, generating a key, encrypting and decrypting */
+    int hpke_mode=OSSL_HPKE_MODE_BASE;
+    OSSL_HPKE_SUITE hpke_suite = OSSL_HPKE_SUITE_DEFAULT;
+    OSSL_HPKE_CTX *ctx = NULL, *rctx = NULL;
+    size_t publen=OSSL_HPKE_MAXSIZE; unsigned char pub[OSSL_HPKE_MAXSIZE];
```

Comment:
Not sure we should be relying on this OSSL_HPKE_MAXSIZE define.. Smaller buffers can be used.

Reply:
fixed